File: server/src/main/java/neuroevolution/math/Matrix.java
package neuroevolution.math;

import neuroevolution.exception.InvalidDimensionsException;

public class Matrix {

	private final int height;
	private final int width;
	private final double[][] elements;

	public Matrix(final int height, final int width, final MatrixValueProvider valueProvider) {
		this.height = height;
		this.width = width;
		elements = new double[height][width];
		for (int row = 0; row < height; ++row) {
			for (int column = 0; column < width; ++column) {
				elements[row][column] = valueProvider.generateValue(row, column);
			}
		}
	}

	public int height() {
		return height;
	}

	public int width() {
		return width;
	}

	public double get(final int row, final int column) {
		if (row < 0 || row >= height || column < 0 || column >= width) {
			throw new IndexOutOfBoundsException("Cannot access matrix element. Index out of bound.");
		}

		return elements[row][column];
	}

	public Vector multiply(final Vector vector) {
		if (width != vector.length()) {
			throw new InvalidDimensionsException("Cannot multiply. Matrix and vector dimensions does not match.");
		}

		final VectorValueProvider valueProvider = index -> {
			double result = 0;
			for (int i = 0; i < width; ++i) {
				result += elements[index][i] * vector.get(i);
			}
			return result;
		};

		return new Vector(height, valueProvider);
	}

	public Matrix apply(final ValueModifier modifier) {
		return new Matrix(height, width, (row, column) -> modifier.apply(elements[row][column]));
	}

}

File: webapp/src/app/evolution/logic/neural-network/matrix/matrix.ts
import {Vector, VectorElementProvider} from '../vector/vector';

export type MatrixElementProvider = (row: number, column: number) => number;

export class Matrix {


  private constructor(public readonly height: number,
                      public readonly width: number,
                      public readonly elements: readonly (readonly number[])[]) {
  }

  static fromProvider(height: number, width: number, provider: MatrixElementProvider): Matrix {
    const elements: number[][] = [];
    for (let r = 0; r < height; ++r) {
      const row: number[] = [];
      for (let c = 0; c < width; ++c) {
        const element: number = provider(r, c);
        if (!element && element !== 0) {
          throw new Error('Matrix element cannot be empty.');
        }
        row.push(element);
      }
      elements.push(row);
    }
    return new Matrix(height, width, elements);
  }

  multiply(vector: Vector): Vector {
    if (this.width !== vector.length) {
      throw new Error('Cannot multiply matrix by vector. Dimensions does not match.');
    }

    const provider: VectorElementProvider = (index: number): number => {
      let result = 0;
      for (let i = 0; i < this.width; ++i) {
        result += this.elements[index][i] * vector.elements[i];
      }
      return result;
    };

    return Vector.fromProvider(this.height, provider);
  }



}
